src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,752 @@
+
+    // We are canonicalizing and reinserting the shape for other, so remove it
+    // from the map under its current shape.
+    groupShapeIndices.erase(it);
+
+    // Put the types for both groups in to the canonical order.
+    auto permutation = getCanonicalPermutation(groupInfo.group);
+    groupInfo.permutation = otherInfo.permutation = std::move(permutation);
+
+    // Set up `other` to be the representative element of the equivalence class.
+    otherInfo.classInfo.emplace(otherInfo);
+    auto& classInfo = *otherInfo.classInfo;
+
+    // Update both groups to have the initial valid order. Their shapes still
+    // match.
+    classInfo.permute(otherInfo);
+    classInfo.permute(groupInfo);
+
+    // Insert `other` with its new shape. It may end up being joined to another
+    // existing equivalence class, in which case its class info will be cleared
+    // and `group` will subsequently be added to that same existing class, or
+    // alternatively it may be inserted as the representative element of a new
+    // class to which `group` will subsequently be joined.
+    shapesToUpdate.push_back(group);
+    shapesToUpdate.push_back(other);
+  }
+
+  std::vector<size_t>
+  getCanonicalPermutation(const std::vector<HeapType>& types) {
+    // The correctness of this part depends on some interesting properties of
+    // strongly connected graphs with ordered, directed edges. A permutation of
+    // the vertices in a graph is an isomorphism that produces an isomorphic
+    // graph. An automorphism is an isomorphism of a structure onto itself, so
+    // an automorphism of a graph is a permutation of the vertices that does not
+    // change the graph. Permutations can be described in terms of sets of
+    // cycles of elements.
+    //
+    // Theorem 1: All cycles in an automorphism of an SCC are the same size.
+    //
+    //     Proof: By contradiction. Assume there are two cycles of different
+    //     sizes. Because the SCC is fully connected, there is a path from an
+    //     element in the smaller of these cycles to an element in the larger.
+    //     This path must contain an edge between two edges in cycles of
+    //     different sizes N and M such that N < M. Apply the automorphism N
+    //     times. The source of this edge has been cycled back to its original
+    //     index, but the destination is not yet back at its original index, so
+    //     the edge's destination index is different from its original
+    //     destination index and the permutation we applied was not an
+    //     automorphism after all.
+    //
+    // Corollary 1.1: No nontrivial automorphism of an SCC may have a stationary
+    // element, since either all the cycles have size 1 and the automorphism is
+    // trivial or all the cycles have some other size and there are no
+    // stationary elements.
+    //
+    // Corollary 1.2: All orderings with the same first element are distinct
```

Comment:
What does "ordering" mean here? Isomorphism? Automorphism?

I am guessing automorphism since we just proved something about automorphisms. But if so then I don't see what 1.2 is true. Yes, no nontrivial automorphism can keep the first element stationary, but that doesn't seem to directly imply anything about multiple automorphisms with no stationary elements whatsoever that happen to have identical first elements.

Reply:
This could be rewritten "No two isomorphic strongly connected type graphs with the same first element are automorphisms of each other."

Reply:
i.e. an "ordering" (of types) is a particular isomorphism of a type graph.